api/apps/api/src/modules/projects/blm/change-blm-range.handler.ts

```diff
@@ -1,18 +1,82 @@
 import { CommandHandler, IInferredCommandHandler } from '@nestjs/cqrs';
-import { Either, right } from 'fp-ts/Either';
+import { Either, isLeft, left, right } from 'fp-ts/Either';
 
-import { ChangeBlmRange, ChangeRangeErrors } from './change-blm-range.command';
+import {
+  ChangeBlmRange,
+  ChangeRangeErrors,
+  queryFailure,
+} from './change-blm-range.command';
 import { ProjectBlmRepo } from '@marxan-api/modules/blm';
+import { InjectEntityManager, InjectRepository } from '@nestjs/typeorm';
+import { Project } from '@marxan-api/modules/projects/project.api.entity';
+import { EntityManager, Repository } from 'typeorm';
+import { DbConnections } from '@marxan-api/ormconfig.connections';
+import { Logger } from '@nestjs/common';
+import { SetProjectBlm } from '@marxan-api/modules/projects/blm/set-project-blm';
+import { BlmValuesCalculator } from '@marxan-api/modules/projects/blm/domain/blm-values-calculator';
 
 @CommandHandler(ChangeBlmRange)
 export class ChangeBlmRangeHandler
   implements IInferredCommandHandler<ChangeBlmRange> {
-  constructor(private readonly blmRepository: ProjectBlmRepo) {}
+  private readonly logger: Logger = new Logger(SetProjectBlm.name);
+
+  constructor(
+    @InjectRepository(Project)
+    protected readonly projectRepository: Repository<Project>,
+    @InjectEntityManager(DbConnections.geoprocessingDB)
+    private readonly entityManager: EntityManager,
+    private readonly blmRepository: ProjectBlmRepo,
+  ) {}
 
   async execute({
     projectId,
     range,
   }: ChangeBlmRange): Promise<Either<ChangeRangeErrors, true>> {
+    if (this.isInValidRange(range)) {
+      this.logger.error(
+        `Received range [${range[0]},${range[1]}] for project with ID: ${projectId} is invalid`,
+      );
+
+      return left(queryFailure);
+    }
+
+    const project = await this.projectRepository.findOneOrFail(projectId);
+    const area = await this.getPlanningUnitArea(project);
+
+    if (!area) {
+      this.logger.error(
+        `Could not get Planning Unit area for project with ID: ${projectId}`,
+      );
+
+      return left(queryFailure);
+    }
+
+    const blmValues = BlmValuesCalculator.with(range, area);
+
+    const result = await this.blmRepository.update(projectId, range, blmValues);
+    if (isLeft(result))
+      this.logger.error(
+        `Could not update BLM for project with ID: ${projectId}`,
+      );
+
     return right(true);
   }
+
+  private isInValidRange(range: [number, number]) {
+    return range.length !== 2 && range[0] < 0 && range[0] > range[1];
+  }
```

Comment:
Given how I had to ask colleagues to cross-check unrelated things for me this afternoon, I think it's safe to assume that my brain has already left for the weekend, however, from the checks we're making here and the way the method is used above, should this not actually be as below?

```suggestion
  private isInValidRange(range: [number, number]) {
    return range.length !== 2 || range[0] < 0 || range[0] > range[1];
  }
```

I would actually go a bit further, clarify the naming (contentious issue, I know, so this is just a minor suggestion) and avoiding the use of the negative form (invalid vs valid), thus:

```suggestion
  private isRangeValid(range: [number, number]) {
    return range?.length == 2 && range[0] > 0 && range[0] < range[1];
  }
```
This would however require inverting any checks on the return value in consumers of the method - but in general, and this is bordering on personal preference of course, I suggest to check the positive form of something (is this valid?) vs the negative (is this invalid?), as possible uses such as `!isInvalid()` quickly become harder to mentally parse because of double negations, etc.

Reply:
Yep, solved, it was old garbage that I didn't realize because the first validation is done in the DTO, but already fixed